File: interviewbit-arrays-max-sum-contiguous-subarray.java
public class Solution {
    // DO NOT MODFIY THE LIST.
    public int maxSubArray(final List<Integer> a) {
        int maxSoFar = a.get(0);
        int maxEndingHere = a.get(0);
        for (int i = 1; i < a.size(); i++) {
            maxEndingHere = Math.max(maxEndingHere + a.get(i), a.get(i));
            maxSoFar = Math.max(maxSoFar, maxEndingHere);
        }
        return maxSoFar;
    }
}

File: interviewbit-arrays-max-sum-contiguous-subarray.py
class Solution:
    # @param A : tuple of integers
    # @return an integer
    def maxSubArray(self, A):
        max_ending_here = max_so_far = A[0]
        for ele in A[1:]:
            max_ending_here = max(max_ending_here + ele, ele)
            max_so_far = max(max_so_far, max_ending_here)

        return max_so_far
